src/app/penalties/penalties.ts

```diff
@@ -119,27 +125,67 @@ export class PenaltiesComponent implements OnInit {
   }
 
   onPageClicked(pageNumber) {
-    if (pageNumber < 1) {
-      pageNumber = 1;
-    } else if (pageNumber > this.pages.length) {
-      pageNumber = this.pages.length;
-    }
-    if (pageNumber !== this.activePage) {
-      this.activePage = pageNumber;
-      this.getPenalties().publish().connect();
+    if (!this.isValidPage(pageNumber)) {
+      return;
     }
+
+    this.activePage = pageNumber;
+    this.getPenalties()
+      .publish()
+      .connect();
+  }
+
+  isValidPage(page: number): boolean {
+    let isPositive,
+        isLessThatMax;
+
+    isPositive = 1 <= page;
+    isLessThatMax = page <= _.last(this.pages);
+
+    return isPositive && isLessThatMax;
+  }
+
+  createPaginationInfo(res) {
+    let visiblePages,
+        adjustBy;
+
+    visiblePages = _.range(
+      this.activePage - 2,
+      this.activePage + 3
+    );
+
+    adjustBy = visiblePages[ 0 ] < 1
+      ? Math.abs(visiblePages[ 0 ]) + 1
+      : 0;
+
+    visiblePages = visiblePages
+      .map(page => page + adjustBy)
+      .filter(page => this.isValidPage(page));
+
+    this.canGoToPreviousPage = _.first(this.pages) !== this.activePage;
+    this.canGoToNextPage = _.last(this.pages) !== this.activePage;
+
+    this.visiblePages = visiblePages;
   }
 
   getPenalties() {
     return this.http.get('/api/penalties?page=' + this.activePage)
       .catch(err => this.logError(err))
       .map(res => {
         res = res.json();
-        this.pages = Array(res.pagination.pages).fill(0).map((x, i) => i + 1);
+
         this.penalties = res.objects;
+
+        this.pages = Array(res.pagination.pages)
+          .fill(0)
+          .map((x, i) => i + 1);
+
         if (this.penalties.length > 0) {
           this.setActivePenalty(this.penalties[0]);
         }
+
+        this.createPaginationInfo(res);
```

Comment:
Since you moved logic to separate component, this could be performed elsewhere -
maybe `ngOnChanges` of `PenaltiesPagination`.